File: go/signer.go
package remitmd

import (
	"crypto/ecdsa"
	"encoding/hex"
	"fmt"
	"math/big"
	"strings"

	"github.com/ethereum/go-ethereum/accounts/abi"
	"github.com/ethereum/go-ethereum/common"
	"github.com/ethereum/go-ethereum/crypto"
)


// Signer signs EIP-712 typed data for authenticating API requests.
// Implement this interface to use hardware wallets, KMS, or other signing backends.
type Signer interface {
	// Sign returns an EIP-712 signature over the given digest (32 bytes).
	Sign(digest [32]byte) ([]byte, error)
	// SignHash signs a raw 32-byte hash and returns the 0x-prefixed hex signature (65 bytes: r+s+v).
	// Used by server-side permit/x402 flows where the server computes the EIP-712 hash.
	SignHash(hash []byte) (string, error)
	// Address returns the Ethereum address of the signing key.
	Address() common.Address
}

// PrivateKeySigner implements Signer using an in-memory ECDSA private key.
// For production agents, prefer a KMS-backed signer.
type PrivateKeySigner struct {
	key     *ecdsa.PrivateKey
	address common.Address
}

// NewPrivateKeySigner creates a Signer from a hex-encoded private key.
// The key may be 0x-prefixed or bare hex.
func NewPrivateKeySigner(hexKey string) (*PrivateKeySigner, error) {
	hexKey = strings.TrimPrefix(hexKey, "0x")
	keyBytes, err := hex.DecodeString(hexKey)
	if err != nil {
		return nil, remitErr(ErrCodeUnauthorized, "invalid private key: expected hex-encoded 32-byte value", map[string]any{
			"hint": "private keys are 64 hex characters, optionally prefixed with 0x",
		})
	}
	key, err := crypto.ToECDSA(keyBytes)
	if err != nil {
		return nil, remitErr(ErrCodeUnauthorized, fmt.Sprintf("invalid private key: %s", err), nil)
	}
	return &PrivateKeySigner{
		key:     key,
		address: crypto.PubkeyToAddress(key.PublicKey),
	}, nil
}

// Sign produces an Ethereum signature (65 bytes: r, s, v) over a 32-byte digest.
func (s *PrivateKeySigner) Sign(digest [32]byte) ([]byte, error) {
	sig, err := crypto.Sign(digest[:], s.key)
	if err != nil {
		return nil, fmt.Errorf("signing failed: %w", err)
	}
	// go-ethereum returns v as 0 or 1; Ethereum expects 27 or 28
	sig[64] += 27
	return sig, nil
}

// SignHash signs a raw 32-byte hash and returns a 0x-prefixed hex signature string.
func (s *PrivateKeySigner) SignHash(hash []byte) (string, error) {
	if len(hash) != 32 {
		return "", fmt.Errorf("SignHash: expected 32 bytes, got %d", len(hash))
	}
	var digest [32]byte
	copy(digest[:], hash)
	sig, err := s.Sign(digest)
	if err != nil {
		return "", err
	}
	return "0x" + hex.EncodeToString(sig), nil
}

// Address returns the Ethereum address for this signing key.
func (s *PrivateKeySigner) Address() common.Address {
	return s.address
}

// ─── EIP-712 domain and API request authentication ─────────────────────────────

var (
	domainType = abi.Arguments{
		{Type: mustType("bytes32")},  // typeHash
		{Type: mustType("bytes32")},  // name hash
		{Type: mustType("bytes32")},  // version hash
		{Type: mustType("uint256")},  // chainId
		{Type: mustType("address")},  // verifyingContract
	}
	apiRequestType = abi.Arguments{
		{Type: mustType("bytes32")},  // typeHash
		{Type: mustType("bytes32")},  // keccak256(method)
		{Type: mustType("bytes32")},  // keccak256(path)
		{Type: mustType("uint256")},  // timestamp
		{Type: mustType("bytes32")},  // nonce
	}

	domainTypeHash      = crypto.Keccak256Hash([]byte("EIP712Domain(string name,string version,uint256 chainId,address verifyingContract)"))
	apiRequestTypeHash  = crypto.Keccak256Hash([]byte("APIRequest(string method,string path,uint256 timestamp,bytes32 nonce)"))
	nameHash            = crypto.Keccak256Hash([]byte("remit.md"))
	versionHash         = crypto.Keccak256Hash([]byte("0.1"))
)

// computeDomainSeparator returns the EIP-712 domain separator for a given chain and contract.
func computeDomainSeparator(chainID *big.Int, contract common.Address) [32]byte {
	packed, _ := domainType.Pack(domainTypeHash, nameHash, versionHash, chainID, contract)
	return crypto.Keccak256Hash(packed)
}

// computeRequestDigest computes the EIP-712 digest for authenticating an API request.
//
// This matches the server's auth middleware (auth.rs: compute_eip712_hash).
// Struct: APIRequest(string method, string path, uint256 timestamp, bytes32 nonce)
func computeRequestDigest(chainID *big.Int, contract common.Address, method, path string, timestamp uint64, nonce [32]byte) [32]byte {
	domain := computeDomainSeparator(chainID, contract)

	methodHash := crypto.Keccak256Hash([]byte(method))
	pathHash := crypto.Keccak256Hash([]byte(path))

	packed, _ := apiRequestType.Pack(
		apiRequestTypeHash,
		methodHash,
		pathHash,
		new(big.Int).SetUint64(timestamp),
		nonce,
	)
	structHash := crypto.Keccak256Hash(packed)

	return crypto.Keccak256Hash(
		append([]byte("\x19\x01"), append(domain[:], structHash[:]...)...),
	)
}

func mustType(t string) abi.Type {
	typ, err := abi.NewType(t, "", nil)
	if err != nil {
		panic(fmt.Sprintf("abi.NewType(%q): %v", t, err))
	}
	return typ
}

File: rust/src/signer.rs
use k256::ecdsa::{signature::hazmat::PrehashSigner, SigningKey};
use sha3::{Digest, Keccak256};

use crate::error::{codes, remit_err, remit_err_ctx, RemitError};

/// Signs EIP-712 typed data for authenticating API requests.
///
/// Implement this trait to use hardware wallets, KMS, or other signing backends.
/// For most agents, `PrivateKeySigner` is sufficient.
pub trait Signer: Send + Sync {
    /// Sign a 32-byte digest and return a 65-byte Ethereum signature (r, s, v).
    fn sign(&self, digest: &[u8; 32]) -> Result<Vec<u8>, RemitError>;

    /// Sign a 32-byte hash and return the 0x-prefixed hex signature (65 bytes: r+s+v).
    ///
    /// Used by `/permits/prepare` and `/x402/prepare` flows where the server
    /// computes the EIP-712 hash and the SDK only signs it.
    fn sign_hash(&self, hash: &[u8; 32]) -> Result<String, RemitError> {
        let sig_bytes = self.sign(hash)?;
        Ok(format!("0x{}", hex::encode(&sig_bytes)))
    }

    /// Return the Ethereum address (checksummed hex, `0x`-prefixed) for this key.
    fn address(&self) -> &str;
}

/// Signs API requests using an in-memory ECDSA private key.
///
/// The private key is held in memory and never logged or included in error messages.
/// For production agents, prefer a KMS-backed signer.
///
/// ```rust,ignore
/// use remitmd::PrivateKeySigner;
///
/// let signer = PrivateKeySigner::new("0xdeadbeef...")?;
/// println!("address: {}", signer.address());
/// ```
pub struct PrivateKeySigner {
    signing_key: SigningKey,
    address: String,
}

impl PrivateKeySigner {
    /// Create a signer from a hex-encoded private key (with or without `0x` prefix).
    pub fn new(hex_key: &str) -> Result<Self, RemitError> {
        let hex_key = hex_key.trim_start_matches("0x");
        let key_bytes = hex::decode(hex_key).map_err(|_| {
            remit_err_ctx(
                codes::INVALID_KEY,
                "invalid private key: expected 64 hex characters (32 bytes), optionally prefixed with 0x",
                "hint",
                "private keys are 64 hex characters, optionally prefixed with 0x",
            )
        })?;

        let signing_key = SigningKey::from_bytes(key_bytes.as_slice().into()).map_err(|e| {
            remit_err(
                codes::INVALID_KEY,
                format!("invalid private key: {e}. Key must be a valid 32-byte secp256k1 scalar."),
            )
        })?;

        let address = ethereum_address(signing_key.verifying_key());
        Ok(Self {
            signing_key,
            address,
        })
    }
}

impl Signer for PrivateKeySigner {
    fn sign(&self, digest: &[u8; 32]) -> Result<Vec<u8>, RemitError> {
        let (sig, recovery_id) = self
            .signing_key
            .sign_prehash(digest)
            .map_err(|e| remit_err(codes::INVALID_SIGNATURE, format!("signing failed: {e}")))?;

        let mut bytes = sig.to_bytes().to_vec(); // 64 bytes (r, s)
                                                 // Convert recovery_id to Ethereum's v (27 or 28)
        bytes.push(recovery_id.to_byte() + 27);
        Ok(bytes)
    }

    fn address(&self) -> &str {
        &self.address
    }
}

/// Derive the checksummed Ethereum address from a k256 verifying key.
fn ethereum_address(verifying_key: &k256::ecdsa::VerifyingKey) -> String {
    // Uncompressed public key: 0x04 || x (32) || y (32) = 65 bytes total
    let point = verifying_key.to_encoded_point(false);
    let pub_bytes = point.as_bytes();

    // Hash the 64-byte x||y (skipping 0x04 prefix)
    let mut hasher = Keccak256::new();
    hasher.update(&pub_bytes[1..]);
    let hash = hasher.finalize();

    // Last 20 bytes of the hash = Ethereum address
    let addr_bytes = &hash[12..];
    format!("0x{}", hex::encode(addr_bytes))
}
